### PR TITLE
GH-92239: Respect PEP 523

### PR DESCRIPTION
**Pull request by markshannon**
Make sure that we respect PEP 523 when specializing

### CODE REVIEW

**Comment by markshannon** [comment]
@itamaro

File: Python/specialize.c [hunk 1]
```diff
@@ -1466,6 +1467,11 @@ specialize_py_call(PyFunctionObject *func, _Py_CODEUNIT *instr, int nargs,
     assert(_Py_OPCODE(*instr) == CALL_ADAPTIVE);
     PyCodeObject *code = (PyCodeObject *)func->func_code;
```
Comment (itamaro) [inline Python/specialize.c:1468]: will we want to add a check here for overridden `vectorcall` on `func` as well, or is it already covered somewhere else? (assuming the vectorcall set API sets the version to zero)

File: Lib/test/test_capi.py [hunk 2]
```diff
@@ -1142,5 +1142,33 @@ def test_frame_get_generator(self):
...
+        def func1():
+            pass
+        self.do_test(func1)
+
+    def test_specialize_before_intercept(self):
+        def func2():
+            pass
+        for _ in range(SUFFICIENT_TO_SPECIALIZE):
+            func2()
+        self.do_test(func2)
```
Comment (itamaro) [inline Lib/test/test_capi.py]: more for my understanding - this test case fails without the added PEP-523 checks? (but the other one passes?)

File: Python/ceval.c [hunk 3]
```diff
@@ -4890,6 +4890,7 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
 
         TARGET(CALL_PY_EXACT_ARGS) {
             assert(call_shape.kwnames == NULL);
+            DEOPT_IF(tstate->interp->eval_frame, CALL);
```
Comment (itamaro) [inline Python/ceval.c:4893]: I don't know how to verify that these opcodes (`CALL_PY_EXACT_ARGS` & `CALL_PY_WITH_DEFAULTS`) are the only ones where this check is needed - I trust you :)

**Review by itamaro — COMMENTED** [review]
lgtm, thanks @markshannon !

File: Python/specialize.c (hunk 1 above)
Comment (markshannon) [reply Python/specialize.c:1468]: I think we want to check for that in https://github.com/python/cpython/pull/92257

File: Lib/test/test_capi.py (hunk 2 above)
Comment (markshannon) [reply Lib/test/test_capi.py]: Both tests should fail without the checks.

File: Python/ceval.c (hunk 3 above)
Comment (markshannon) [reply Python/ceval.c:4893]: Specializations of `CALL` and `PRECALL` listed [here](https://github.com/python/cpython/blob/main/Lib/opcode.py#L256) and [here](https://github.com/python/cpython/blob/main/Lib/opcode.py#L300)

**Pull request merged**